golem-worker-service/src/api/api_deployment.rs

```diff
@@ -182,4 +183,102 @@ impl ApiDeploymentApi {
+    #[oai(
+        path = "/:site/:id/:version",
+        method = "delete",
+        operation_id = "undeploy_api"
+    )]
+    async fn undeploy_api(
+        &self,
+        site: Path<String>,
+        id: Path<String>,
+        version: Path<String>,
+    ) -> Result<Json<String>, ApiEndpointError> {
+        let record = recorded_http_api_request!(
+            "undeploy_api",
+            site = site.0.clone(),
+            id = id.0.clone(),
+            version = version.0.clone()
+        );
+
+        let response = self
+            .undeploy_api_internal(site.0, id.0, version.0, &EmptyAuthCtx::default())
+            .instrument(record.span.clone())
+            .await;
+        record.result(response)
+    }
+
+    async fn undeploy_api_internal(
+        &self,
+        site: String,
+        id: String,
+        version: String,
+        auth_ctx: &EmptyAuthCtx,
+    ) -> Result<Json<String>, ApiEndpointError> {
+        let namespace = DefaultNamespace::default();
+        let api_definition_key = ApiDefinitionIdWithVersion {
+            id: ApiDefinitionId(id),
+            version: ApiVersion(version),
+        };
+
+        // Split the site string into host and subdomain parts
+        let (host, subdomain) = if let Some(idx) = site.find('.') {
+            let (subdomain, host) = site.split_at(idx);
+            (
+                host.trim_start_matches('.').to_string(),
+                Some(subdomain.to_string()),
+            )
+        } else {
+            (site, None)
+        };
+
+        let api_site = ApiSite { host, subdomain };
+
+        // Check if the site exists
+        let site_exists = self
+            .deployment_service
+            .get_by_site(&ApiSiteString::from(&api_site))
+            .await?
+            .is_some();
+
+        if !site_exists {
+            return Err(ApiEndpointError::not_found(safe(
+                "Site not found".to_string(),
+            )));
+        }
+
+        // Check if the API definition exists
+        let api_definition_exists = self
+            .deployment_service
+            .get_by_id(&namespace, Some(api_definition_key.id.clone()))
+            .await?
+            .iter()
+            .any(|deployment| {
+                deployment.api_definition_keys.iter().any(|key| {
+                    key.id == api_definition_key.id && key.version == api_definition_key.version
+                })
+            });
+
+        if !api_definition_exists {
+            return Err(ApiEndpointError::not_found(safe(
+                "API definition not found".to_string(),
+            )));
+        }
+
+        let api_deployment = gateway_api_deployment::ApiDeploymentRequest {
+            namespace: namespace.clone(),
+            api_definition_keys: vec![api_definition_key],
+            site: api_site,
+        };
+
+        self.deployment_service
+            .undeploy(&api_deployment, auth_ctx)
+            .await?;
+
```

Comment:
Please move this logic out of the route into the ApiDeploymentService. Try to only do simple data transformations here.
The new function in the ApiDeploymentService should take a generic AuthCtx as a parameter, the concrete EmptyAuthCtx should be provided to the function here when calling it. 

Looks good otherwise, thanks!